ambry-commons/src/main/java/com/github/ambry/compression/CompressionFactory.java

```diff
@@ -0,0 +1,130 @@
+      }
+    }
+
+    return defaultFactory;
+  }
+
+  // Map from algorithm name to instance of compression service.
+  private final Map<String, Compression> registeredCompressionMap = new HashMap<>();
+
+  // The default compressor to use to compress new data.
+  private Compression defaultCompressor;
+
+  /**
+   * Register a compression service.  Every compression service has a unique name.
+   * Decompression will use the registered compression services to find the decompression algorithm.
+   *
+   * @param newCompressionService The compression service to register.
+   */
+  public void register(Compression newCompressionService) {
+    if (newCompressionService == null) {
+      throw new IllegalArgumentException("newCompressionService");
+    }
+    registeredCompressionMap.put(newCompressionService.getAlgorithmName(), newCompressionService);
+  }
+
+  /**
+   * To get a compressor by algorithm name.
+   * The application config could call this method to set the default compressor algorithm.
+   *      CompressionFactory factory = CompressionFactory.getDefault();
+   *      Compression compressor = factory.getCompressorByName("LZ4");
+   *      factory.setDefaultCompressor(compressor);
+   *
+   * @param algorithmName Name of the algorithm to get.
+   * @return Instance of the compression algorithm; null if not found.
+   */
+  public Compression getCompressorByName(String algorithmName) {
+    return registeredCompressionMap.get(algorithmName);
+  }
+
+  /**
+   * Set the default compressor to use when compressing data.  The new compressor can be null to clear default.
+   * If a new compressor is specified, it must be registered.
+   * @param newCompressor The new compressor to use.
+   */
+  public void setDefaultCompressor(Compression newCompressor) {
+    if (newCompressor != null && !registeredCompressionMap.containsKey(newCompressor.getAlgorithmName())) {
+      throw new IllegalArgumentException("newCompressor is not registered.");
+    }
+    defaultCompressor = newCompressor;
+  }
+
+  /**
+   * Get the default compressor to use when compressing data.
+   * It returns the compressor set by setDefaultCompressor().
+   * @return The default compressor.  Return null if not set.
+   */
+  public Compression getDefaultCompressor() {
+    return defaultCompressor;
+  }
+
+  /**
+   * Given the compressed buffer generated by compressor's compress() method,
+   * find the decompressor based on the algorithm name stored in compressedBuffer.
+   *
+   * @param compressedBuffer The compressed buffer generated by the compress() method.
+   * @return The decompressor that can decompress this compressed buffer.  Null if no decompressor found.
+   */
+  public Compression getCompressionFromCompressedData(byte[] compressedBuffer) {
+    // Compressed buffer contains at least version (1 byte), algorithm name size (1 byte) and original size (4 bytes).
+    if (compressedBuffer == null || compressedBuffer.length < 6) {
```

Comment:
the length check here could be header version dependent.

Reply:
This is related to the 6 check above.  This size 6 check is optional.  The size 6 is minimal that should be applicable to all versions.  When we add a new version, the change can revisit this minimal requirement.